integration/shareddata/iterator.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shareddata
+
+import (
+	"sync"
+
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/resource"
+)
+
+// newValuesIterator creates iterator that iterates through bson.D documents generated
+// by generator function.
+// Generator should return next bson.D document on every execution.
+// To stop iterator, generator must return nil.
+func newValuesIterator(generator func() bson.D) iterator.Interface[struct{}, bson.D] {
+	iter := &valuesIterator{
+		token:     resource.NewToken(),
+		generator: generator,
+	}
+
+	resource.Track(iter, iter.token)
+
+	return iter
+}
+
+// valuesIterator iterates through bson.D documents generated by provided generator function.
+type valuesIterator struct {
+	generator func() bson.D
+	token     *resource.Token
+	m         sync.Mutex
```

Comment:
Mutex should typically be on top, like a hat: https://dmitri.shuralyov.com/idiomatic-go/entries/2

If there is some reason for it being there (alignment, for example), that should be documented